### ISSUE TITLE
training issues

### ISSUE THREAD

**Issue by asmodaay** [NVIDIA/flowtron#12]
Hi, i tried to train model with only LJ data, and with only own data, with fp16 and with fr32, with 1 gpu and with 3 gpu, but everywhere i have this 
![Снимок экрана 2020-05-17 в 19 46 32](https://user-images.githubusercontent.com/45363107/82154603-af6e6800-9877-11ea-8a7f-4ade1bcbc82f.png)
Always los is Nan.
When i start with pretrained chekpoint your code return this:
![Снимок экрана 2020-05-17 в 19 52 44](https://user-images.githubusercontent.com/45363107/82154657-03794c80-9878-11ea-97e0-eaa6919fbc84.png)
I solve it by changing `def load_checkpoint` , but loss is nan(

do u have any ideas what am i doing wrong?

**Comment by wietsedv** [comment]
It does not work properly with the latest PyTorch version (I have not searched for exact problems yet). PyTorch 1.0.0 works for me. I have some edge case problems with 1.0.1 so I would install 1.0.0.

**Comment by asmodaay** [comment]
ty, i will try!

**Comment by rafaelvalle** [comment]
replace .byte() with .bool()
https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

**Comment by asmodaay** [comment]
> replace .byte() with .bool()
> https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

This helps me, and i got this attention plot, but inference samples on this speaker doesnt have any  meaningful speech. Is attention ok?
![Снимок экрана 2020-05-18 в 15 24 30](https://user-images.githubusercontent.com/45363107/82212970-19404d80-991c-11ea-8117-afdc95dc7a6f.png)

**Comment by rafaelvalle** [comment]
This attention plot looks really good! 
Can you show both attention layers? This is only layer 1.

**Comment by asmodaay** [comment]
> This attention plot looks really good!
> Can you show both attention layers? This is only layer 1.

![image](https://user-images.githubusercontent.com/45363107/82241297-91bb0480-9944-11ea-8514-8156d38978ae.png)

**Comment by rafaelvalle** [comment]
Both alignment plots look good.
Can you perform inference on your speaker and share the alignment plots and audio?

**Comment by asmodaay** [comment]
![lj](https://user-images.githubusercontent.com/45363107/82254399-71e30b00-995b-11ea-9088-562035725db9.png)
![pkd](https://user-images.githubusercontent.com/45363107/82254415-77405580-995b-11ea-8209-723f894af8ee.png)

[samples.zip](https://github.com/NVIDIA/flowtron/files/4646411/samples.zip)
Here is 2 sample with lj speaker and another

**Comment by rafaelvalle** [comment]
Looks like the model hasn't learned to attend on your speaker.

**Comment by asmodaay** [comment]
> Looks like the model hasn't learned to attend on your speaker.

Do u have any ideas how to fix it? more data or smth like that, ill try to add my speaker in libri group and finetune lj checkpoint

**Comment by nyanye** [comment]
![image](https://user-images.githubusercontent.com/21135812/82297562-c931bb80-99ed-11ea-854b-72114da0d57c.png)
I'm currently training my own data (~2hrs) from LibriTTS pretrained model and looks like the model hasn't learned to attend yet too.
This is my configuration.
```
{
    "train_config": {
        "output_directory": "/some_output_directory/",
        "epochs": 10000000,
        "learning_rate": 1e-4,
        "weight_decay": 1e-6,
        "sigma": 1.0,
        "iters_per_checkpoint": 5000,
        "batch_size": 4,
        "seed": 1234,
        "checkpoint_path": "/some_checkpoint_path/",
	    "ignore_layers": [],
        "include_layers": ["speaker", "encoder", "embedding"],
        "warmstart_checkpoint_path": "",
        "with_tensorboard": true,
        "fp16_run": true
    },
    "data_config": {
        "training_files": "../../path_to_datasets/train_filelist.txt", 
        "validation_files": "../../path_to_datasets/val_filelist.txt",
        "text_cleaners": ["flowtron_cleaners"],
        "p_arpabet": 0.5,
        "cmudict_path": "data/cmudict_dictionary",
        "sampling_rate": 22050,
        "filter_length": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "mel_fmin": 0.0,
        "mel_fmax": 8000.0,
        "max_wav_value": 32768.0
    },
    "dist_config": {
        "dist_backend": "nccl",
        "dist_url": "tcp://localhost:54321"
    },
    
    "model_config": {
        "n_speakers": 1,
        "n_speaker_dim": 128,
        "n_text": 185,
        "n_text_dim": 512,
        "n_flows": 2,
        "n_mel_channels": 80,
        "n_attn_channels": 640,
        "n_hidden": 1024,
        "n_lstm_layers": 2,
        "mel_encoder_n_hidden": 512,
        "n_components": 0,
        "mean_scale": 0.0,
        "fixed_gaussian": true,
        "dummy_speaker_embedding": false,
        "use_gate_layer": true
    } 
}
```

**Comment by rafaelvalle** [comment]
Rule of thumb is to train the model until it starts overfitting or the loss plateaus.

**Comment by nyanye** [comment]
Yeah I have trained several tacotron 2 networks before with my own datasets and I tend to see the trend that since my dataset is relatively small (~1hours or ~2hours) the validation loss shows overfitting in very early stage of training (like 1k steps). However the results generated by the model tends to get better even after overfitting is started and the alignments looked fine.
In this training I found that flowtron model also overfits really fast (like 500 steps) and in that stage of training the model couldn't really reconstruct the features of my datasets. Is there any empirical way to figure this out?

![image](https://user-images.githubusercontent.com/21135812/82388060-93cfb100-9a73-11ea-823a-372c1d5aee29.png)

**Comment by rafaelvalle** [comment]
Combine your speaker with data from other speakers and train again.

**Comment by nyanye** [comment]
I've currently combined my speaker data with libritts-100 and training it from scratch without pretrained model with batch size of 1.
Do you know the empirical steps needed to see the alignments from tensorboard?
And here's my previous model's mel-spectrograms and alignment plots.

![sid0_sigma0 5_attnlayer0](https://user-images.githubusercontent.com/21135812/82527394-4d607c00-9b71-11ea-8987-d05b05f884e7.png)
![sid0_sigma0 5_attnlayer1](https://user-images.githubusercontent.com/21135812/82527399-4f2a3f80-9b71-11ea-9831-248eabdd6f51.png)

You can also hear it, The sentence is "It is well know that deep generative models have a deep latent space!"
[sid0_sigma0.5.wav.zip](https://github.com/NVIDIA/flowtron/files/4660771/sid0_sigma0.5.wav.zip)

**Comment by rafaelvalle** [comment]
Try adding your speaker to the pre-trained LJS and LibriTTS, fine-tune them and evaluate which sounds better on your data by looking at the validation loss on your data.

**Comment by nyanye** [comment]
Thank you for helping me rafael :) I've read your paper again and finally understood that the model requires long amount of time and resource (1000 Epochs for NVIDIA's proprietary datasets and 500 epochs for LibriTTS datasets to adapt) to check the best results and I was too impatient!

**Comment by singhaki** [comment]
> replace .byte() with .bool()
> https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

even after changing, I am getting the same error

**Comment by singhaki** [comment]
> Hi, i tried to train model with only LJ data, and with only own data, with fp16 and with fr32, with 1 gpu and with 3 gpu, but everywhere i have this
> ![Снимок экрана 2020-05-17 в 19 46 32](https://user-images.githubusercontent.com/45363107/82154603-af6e6800-9877-11ea-8a7f-4ade1bcbc82f.png)
> Always los is Nan.
> When i start with pretrained chekpoint your code return this:
> ![Снимок экрана 2020-05-17 в 19 52 44](https://user-images.githubusercontent.com/45363107/82154657-03794c80-9878-11ea-97e0-eaa6919fbc84.png)
> I solve it by changing `def load_checkpoint` , but loss is nan(
> 
> do u have any ideas what am i doing wrong?

how did you solve the key error?

**Comment by CookiePPP** [comment]
@singhaki 
```
    if 'optimizer' in checkpoint_dict.keys(): optimizer.load_state_dict(checkpoint_dict['optimizer'])
    if 'amp' in checkpoint_dict.keys(): amp.load_state_dict(checkpoint_dict['amp'])
    if 'learning_rate' in checkpoint_dict.keys(): learning_rate = checkpoint_dict['learning_rate']
    if 'hparams' in checkpoint_dict.keys(): hparams = checkpoint_dict['hparams']
    if 'best_validation_loss' in checkpoint_dict.keys(): best_validation_loss = checkpoint_dict['best_validation_loss']
    if 'average_loss' in checkpoint_dict.keys(): average_loss = checkpoint_dict['average_loss']
    if 'iteration' in checkpoint_dict.keys(): iteration = checkpoint_dict['iteration']
```

**Comment by asmodaay** [comment]
> > Hi, i tried to train model with only LJ data, and with only own data, with fp16 and with fr32, with 1 gpu and with 3 gpu, but everywhere i have this
> > ![Снимок экрана 2020-05-17 в 19 46 32](https://user-images.githubusercontent.com/45363107/82154603-af6e6800-9877-11ea-8a7f-4ade1bcbc82f.png)
> > Always los is Nan.
> > When i start with pretrained chekpoint your code return this:
> > ![Снимок экрана 2020-05-17 в 19 52 44](https://user-images.githubusercontent.com/45363107/82154657-03794c80-9878-11ea-97e0-eaa6919fbc84.png)
> > I solve it by changing `def load_checkpoint` , but loss is nan(
> > do u have any ideas what am i doing wrong?
> 
> how did you solve the key error? as in pretrained model only state_dict is present and not these
> {'model': model_for_saving,
> 'iteration': iteration,
> 'optimizer': optimizer.state_dict(),
> 'learning_rate': learning_rate}

Also u can use this:
![image](https://user-images.githubusercontent.com/45363107/83161143-b6676700-a110-11ea-8572-cb6e501cced4.png)

**Comment by rafaelvalle** [comment]
@singhaki you need to change this **[line](https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33)** such that the function returns an object of type bool instead of byte.

**Comment by KingStorm** [comment]
![image](https://user-images.githubusercontent.com/8191470/84132638-b2c0d200-aa78-11ea-837b-dc66e409f769.png)

Is it normal for the target gate always being 0? I suppose it would be something like voice activity detection tag: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ....... 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]

**Comment by rafaelvalle** [comment]
There should be one value equal to 1. 
Can you confirm on the data itself? It could be a plotting issue.

**Comment by KingStorm** [comment]
@rafaelvalle Thanks.  Yes, I got the idea now. it was indeed a display issue, since I was working on LJS data and got reasonable voice quality. And some of gate display has that value 1.

**Comment by rafaelvalle** [comment]
@asmodaay adding your speaker to LibriTTS sounds like a good idea. Let us know how it goes.

**Comment by rafaelvalle** [comment]
Closing due to inactivity. Please re-open if needed.

**Comment by blx0102** [comment]
> @asmodaay adding your speaker to LibriTTS sounds like a good idea. let us know how it goes.

@rafaelvalle Sorry for adding comment in the closed issue. I'm just confused how to deal with the speaker embedding when "adding speaker to LibriTTS". Should I just put the speaker embedding into the ignore layers?

**Comment by pneumoman** [comment]
> 
> 
> replace .byte() with .bool()
> https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

If you have to do that change, is #27 required also?

**Comment by serg06** [comment]
> replace .byte() with .bool()
> https://github.com/NVIDIA/flowtron/blob/master/flowtron.py#L33

@rafaelvalle This fixed the issue for me with FP16 disabled, but when I switch to FP16 my training loss becomes NaN again. Can you help please?
